core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1242,8 +1242,10 @@ public UnifyResult apply(UnifyRuleCall call) {
   public static MutableAggregate permute(MutableAggregate aggregate,
       MutableRel input, Mapping mapping) {
     ImmutableBitSet groupSet = Mappings.apply(mapping, aggregate.groupSet);
-    ImmutableList<ImmutableBitSet> groupSets =
+    ImmutableList<ImmutableBitSet> unSortedGroupSets =
         Mappings.apply2(mapping, aggregate.groupSets);
+    ImmutableList<ImmutableBitSet> groupSets = ImmutableList.copyOf(
+        ImmutableBitSet.ORDERING.sortedCopy(unSortedGroupSets));
```

Comment:
Can you incorporate this change into Mappings.apply2? Since it is only used by groupSets mapping.